src/gcovr/merging.py

```diff
@@ -382,15 +411,40 @@ def merge_function(
       - ``options.func_opts.merge_function_use_line_max``
       - ``options.func_opts.separate_function``
     """
-    if left.demangled_name != right.demangled_name:
-        raise AssertionError("Function demangled name must be equal.")
-    if left.name != right.name:
-        raise AssertionError("Function name must be equal.")
+    if left.demangled_name is not None and right.demangled_name is not None:
+        if (
+            not options.func_name_opts.ignore_function_name
+            and left.demangled_name != right.demangled_name
+        ):
+            raise AssertionError(
+                f"Demangled name must be equal, got {left.demangled_name} and {right.demangled_name}."
+            )
+    elif left.demangled_name is not None or right.demangled_name is not None:
+        if right.demangled_name is not None:
+            left.demangled_name = right.demangled_name
+        if not options.func_name_opts.ignore_function_name_single_definition:
+            raise AssertionError(
+                f"Demangled function name {left.demangled_name} only defined in one file."
+            )
+
+    if left.name is not None and right.name is not None:
+        if not options.func_name_opts.ignore_function_name and left.name != right.name:
+            raise AssertionError(
+                f"Mangled name must be equal, got {left.name} and {right.name}."
+            )
+    elif left.name is not None or right.name is not None:
+        if right.name is not None:
+            left.name = right.name
+        if not options.func_name_opts.ignore_function_name_single_definition:
+            raise AssertionError(
```

Comment:
Should all these exceptions also be GcovrMergeAssertionError if I see the use below?